File: backend/repositories/feedback.py
"""Feedback submission repository — database access."""

from __future__ import annotations

from sqlalchemy.orm import Session

from database.models import ProductFeedbackSubmission, ResearchSurveySubmission


def create_research_survey(
    db: Session,
    *,
    professional_role: str,
    specialty: str,
    years_experience: str,
    sources: str,
    sources_other: str,
    papers_needed: str,
    most_time_consuming: str,
    most_time_consuming_other: str,
    biggest_problem: str,
    biggest_problem_other: str,
    trust_level: str,
    trust_reason: str,
) -> ResearchSurveySubmission:
    """Store a research survey submission."""
    submission = ResearchSurveySubmission(
        professional_role=professional_role,
        specialty=specialty,
        years_experience=years_experience,
        sources=sources,
        sources_other=sources_other,
        papers_needed=papers_needed,
        most_time_consuming=most_time_consuming,
        most_time_consuming_other=most_time_consuming_other,
        biggest_problem=biggest_problem,
        biggest_problem_other=biggest_problem_other,
        trust_level=trust_level,
        trust_reason=trust_reason,
    )
    db.add(submission)
    db.commit()
    db.refresh(submission)
    return submission


def list_research_surveys(db: Session, limit: int) -> list[ResearchSurveySubmission]:
    """Return recent research survey submissions."""
    return (
        db.query(ResearchSurveySubmission)
        .order_by(ResearchSurveySubmission.created_at.desc())
        .limit(limit)
        .all()
    )


def create_product_feedback(
    db: Session,
    *,
    overall_rating: int,
    ease_of_use_rating: int,
    search_rating: int | None,
    summary_rating: int | None,
    features_used: str,
    most_useful: str,
    problems_encountered: str,
    improvements: str,
    feature_requests: str,
    would_recommend: str,
    contact_email: str,
) -> ProductFeedbackSubmission:
    """Store a product feedback submission."""
    submission = ProductFeedbackSubmission(
        overall_rating=overall_rating,
        ease_of_use_rating=ease_of_use_rating,
        search_rating=search_rating,
        summary_rating=summary_rating,
        features_used=features_used,
        most_useful=most_useful,
        problems_encountered=problems_encountered,
        improvements=improvements,
        feature_requests=feature_requests,
        would_recommend=would_recommend,
        contact_email=contact_email,
    )
    db.add(submission)
    db.commit()
    db.refresh(submission)
    return submission


def list_product_feedback(db: Session, limit: int) -> list[ProductFeedbackSubmission]:
    """Return recent product feedback submissions."""
    return (
        db.query(ProductFeedbackSubmission)
        .order_by(ProductFeedbackSubmission.created_at.desc())
        .limit(limit)
        .all()
    )

File: lib/api/feedback.ts
import { apiFetch } from './client';

export interface ResearchSurveyPayload {
  professional_role: string;
  specialty: string;
  years_experience: string;
  sources: string[];
  sources_other: string;
  papers_needed: string;
  most_time_consuming: string;
  most_time_consuming_other: string;
  biggest_problem: string;
  biggest_problem_other: string;
  trust_level: string;
  trust_reason: string;
  website: string;
}

export interface ProductFeedbackPayload {
  overall_rating: number;
  ease_of_use_rating: number;
  search_rating: number | null;
  summary_rating: number | null;
  features_used: string[];
  most_useful: string;
  problems_encountered: string;
  improvements: string;
  feature_requests: string;
  would_recommend: string;
  contact_email: string | null;
  website: string;
}

export interface ResearchSurveyResponse {
  id: number;
  professional_role: string;
  specialty: string;
  years_experience: string;
  sources: string[];
  sources_other: string;
  papers_needed: string;
  most_time_consuming: string;
  most_time_consuming_other: string;
  biggest_problem: string;
  biggest_problem_other: string;
  trust_level: string;
  trust_reason: string;
  created_at: string;
}

export interface ProductFeedbackResponse {
  id: number;
  overall_rating: number;
  ease_of_use_rating: number;
  search_rating: number | null;
  summary_rating: number | null;
  features_used: string[];
  most_useful: string;
  problems_encountered: string;
  improvements: string;
  feature_requests: string;
  would_recommend: string;
  contact_email: string;
  created_at: string;
}

export class FeedbackResponsesError extends Error {
  status: number;

  constructor(message: string, status: number) {
    super(message);
    this.name = 'FeedbackResponsesError';
    this.status = status;
  }
}

export async function submitResearchSurvey(payload: ResearchSurveyPayload) {
  const res = await apiFetch('feedback/research-methods', {
    method: 'POST',
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    throw new Error(
      (await res.json().catch(() => null))?.detail || 'Could not submit the survey',
    );
  }
  return res.json();
}

export async function submitProductFeedback(payload: ProductFeedbackPayload) {
  const res = await apiFetch('feedback/product', {
    method: 'POST',
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    throw new Error(
      (await res.json().catch(() => null))?.detail || 'Could not submit your feedback',
    );
  }
  return res.json();
}

export async function fetchFeedbackResponses(limit = 500): Promise<{
  research: ResearchSurveyResponse[];
  product: ProductFeedbackResponse[];
}> {
  const safeLimit = Math.min(Math.max(Math.trunc(limit), 1), 500);
  const [researchResponse, productResponse] = await Promise.all([
    apiFetch(`feedback/research-methods?limit=${safeLimit}`),
    apiFetch(`feedback/product?limit=${safeLimit}`),
  ]);

  const failedResponse = [researchResponse, productResponse].find((response) => !response.ok);
  if (failedResponse) {
    const fallback = failedResponse.status === 403
      ? 'Administrator access is required to view responses.'
      : failedResponse.status === 401
        ? 'Sign in with an administrator account to view responses.'
        : 'Could not load survey responses.';
    const payload = await failedResponse.json().catch(() => null);
    throw new FeedbackResponsesError(
      payload?.detail || fallback,
      failedResponse.status,
    );
  }

  return {
    research: await researchResponse.json(),
    product: await productResponse.json(),
  };
}
